[BUGFIX] Skip rootline check in be for records stored at pid 0

rengaw83 opened the pull request:
I got the folowing error in **TYPO3 11** with **PHP 8** after editing a be_user:
```
PHP Warning: Undefined array key "TSFE" in ./typo3conf/ext/solr/Classes/Domain/Index/Queue/RecordMonitor/Helper/RootPageResolver.php line 169
```


This PR fixes an issue in backend for records located on page id 0 like be_users.
If there is no TSFE and the page id is 0, there is no need to check the rootline to get the root page id.

Trace:
```php
[
    0 => './typo3/index.php:{closure}#21',
    1 => './typo3/index.php:TYPO3\\CMS\\Core\\Http\\AbstractApplication->run#20',
    2 => './typo3/sysext/core/Classes/Http/AbstractApplication.php:TYPO3\\CMS\\Backend\\Http\\Application->handle#100',
    3 => './typo3/sysext/backend/Classes/Http/Application.php:TYPO3\\CMS\\Core\\Http\\AbstractApplication->handle#72',
    4 => './typo3/sysext/core/Classes/Http/AbstractApplication.php:TYPO3\\CMS\\Core\\Http\\MiddlewareDispatcher->handle#86',
    5 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#78',
    6 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Core\\Middleware\\VerifyHostHeader->process#172',
    7 => './typo3/sysext/core/Classes/Middleware/VerifyHostHeader.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#55',
    8 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Core\\Middleware\\NormalizedParamsAttribute->process#172',
    9 => './typo3/sysext/core/Classes/Middleware/NormalizedParamsAttribute.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#45',
    10 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:DMK\\MkSanitizedParameters\\Middleware\\GlobalInputSanitizerMiddleware->process#172',
    11 => './typo3conf/ext/mksanitizedparameters/Classes/Middleware/GlobalInputSanitizerMiddleware.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#80',
    12 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Backend\\Middleware\\LockedBackendGuard->process#172',
    13 => './typo3/sysext/backend/Classes/Middleware/LockedBackendGuard.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#75',
    14 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Backend\\Middleware\\ForcedHttpsBackendRedirector->process#172',
    15 => './typo3/sysext/backend/Classes/Middleware/ForcedHttpsBackendRedirector.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#55',
    16 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Backend\\Middleware\\BackendRouteInitialization->process#172',
    17 => './typo3/sysext/backend/Classes/Middleware/BackendRouteInitialization.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#86',
    18 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Backend\\Middleware\\BackendUserAuthenticator->process#172',
    19 => './typo3/sysext/backend/Classes/Middleware/BackendUserAuthenticator.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#165',
    20 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Backend\\Middleware\\OutputCompression->process#172',
    21 => './typo3/sysext/backend/Classes/Middleware/OutputCompression.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#47',
    22 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Backend\\Middleware\\AdditionalResponseHeaders->process#172',
    23 => './typo3/sysext/backend/Classes/Middleware/AdditionalResponseHeaders.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#41',
    24 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Backend\\Middleware\\SiteResolver->process#172',
    25 => './typo3/sysext/backend/Classes/Middleware/SiteResolver.php:Psr\\Http\\Server\\RequestHandlerInterface@anonymous' . "\0" . './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:138$41b->handle#69',
    26 => './typo3/sysext/core/Classes/Http/MiddlewareDispatcher.php:TYPO3\\CMS\\Core\\Middleware\\ResponsePropagation->process#172',
    27 => './typo3/sysext/core/Classes/Middleware/ResponsePropagation.php:TYPO3\\CMS\\Backend\\Http\\RequestHandler->handle#34',
    28 => './typo3/sysext/backend/Classes/Http/RequestHandler.php:TYPO3\\CMS\\Backend\\Http\\RouteDispatcher->dispatch#110',
    29 => './typo3/sysext/backend/Classes/Http/RouteDispatcher.php:TYPO3\\CMS\\Backend\\Controller\\EditDocumentController->mainAction#91',
    30 => './typo3/sysext/backend/Classes/Controller/EditDocumentController.php:TYPO3\\CMS\\Backend\\Controller\\EditDocumentController->processData#411',
    31 => './typo3/sysext/backend/Classes/Controller/EditDocumentController.php:TYPO3\\CMS\\Core\\DataHandling\\DataHandler->process_datamap#563',
    32 => './typo3/sysext/core/Classes/DataHandling/DataHandler.php:TYPO3\\CMS\\Core\\DataHandling\\DataHandler->hook_processDatamap_afterDatabaseOperations#1134',
    33 => './typo3/sysext/core/Classes/DataHandling/DataHandler.php:ApacheSolrForTypo3\\Solr\\IndexQueue\\RecordMonitor->processDatamap_afterDatabaseOperations#781',
    34 => './typo3conf/ext/solr/Classes/IndexQueue/RecordMonitor.php:TYPO3\\CMS\\Core\\EventDispatcher\\EventDispatcher->dispatch#144',
    35 => './typo3/sysext/core/Classes/EventDispatcher/EventDispatcher.php:ApacheSolrForTypo3\\Solr\\Domain\\Index\\Queue\\UpdateHandler\\EventListener\\ImmediateProcessingEventListener->__invoke#51',
    36 => './typo3conf/ext/solr/Classes/Domain/Index/Queue/UpdateHandler/EventListener/ImmediateProcessingEventListener.php:ApacheSolrForTypo3\\Solr\\Domain\\Index\\Queue\\UpdateHandler\\EventListener\\ImmediateProcessingEventListener->handleRecordUpdatedEvent#58',
    37 => './typo3conf/ext/solr/Classes/Domain/Index/Queue/UpdateHandler/EventListener/ImmediateProcessingEventListener.php:ApacheSolrForTypo3\\Solr\\Domain\\Index\\Queue\\UpdateHandler\\DataUpdateHandler->handleRecordUpdate#122',
    38 => './typo3conf/ext/solr/Classes/Domain/Index/Queue/UpdateHandler/DataUpdateHandler.php:ApacheSolrForTypo3\\Solr\\Domain\\Index\\Queue\\UpdateHandler\\DataUpdateHandler->getRecordRootPageIds#229',
    39 => './typo3conf/ext/solr/Classes/Domain/Index/Queue/UpdateHandler/DataUpdateHandler.php:ApacheSolrForTypo3\\Solr\\Domain\\Index\\Queue\\RecordMonitor\\Helper\\RootPageResolver->getResponsibleRootPageIds#371',
    40 => './typo3conf/ext/solr/Classes/Domain/Index/Queue/RecordMonitor/Helper/RootPageResolver.php:ApacheSolrForTypo3\\Solr\\Domain\\Index\\Queue\\RecordMonitor\\Helper\\RootPageResolver->buildResponsibleRootPageIds#94',
    41 => './typo3conf/ext/solr/Classes/Domain/Index/Queue/RecordMonitor/Helper/RootPageResolver.php:ApacheSolrForTypo3\\Solr\\Domain\\Index\\Queue\\RecordMonitor\\Helper\\RootPageResolver->getRootPageIdByTableAndUid#204',
    42 => './typo3conf/ext/solr/Classes/Domain/Index/Queue/RecordMonitor/Helper/RootPageResolver.php:ApacheSolrForTypo3\\Solr\\Domain\\Index\\Queue\\RecordMonitor\\Helper\\RootPageResolver->getRootPageId#235',
];
```

Relates: #3173

dkd-kaehm commented:
@rengaw83 
Thanks for fixing that issue.
There is already a pull-request for same issue #3174, which I commented as well: https://github.com/TYPO3-Solr/ext-solr/pull/3174#issuecomment-1021039093

Would you like to remove the usage of `$GLOBALS['TSFE']` in that class and check the functionality.
All that, because `$GLOBALS['TSFE']` is never initialized on indexing context. That is since https://github.com/TYPO3-Solr/ext-solr/commit/a246cb8e34d80da6227445c0c800620f16264a00 and I didn't touch this class doubtless, because there is no test case on involved methods.

rengaw83 commented:
Hi @dkd-kaehm,

i have removed the tsfe usage now.  
I have removed the forceFallback argument too, it's superfluous in my opinion. The rootline is determined in the same way for fe and be by the typo3 core and its cached in the RootlineUtility of the core.

I have tested the indexing of pages and news in TYPO3 11 with PHP 8 an all works like expected for me.  
Unittests are running too, for intergationtest we have to wait for the github action.

I hope you imagined it that way.

rengaw83 commented:
@dkd-kaehm there are some integration tests failing now.  
i doesn't really under stand it for the moment. just to be sure:
The tests are checking if pages without rootpage are indexed. But in my opinion with typo3 11 and solr 11.5 we are not able to index pages outside a site, so a shared page outside a site is never indexed. Is this correct? Or how can i index a page outside a site at the moment?
If it is possible to index pages outside a site, i has to reimplement the forceFallback argument and fetch the rootline from RootlineUtility only if set. But how to fetch the rootline in fe environment? Can it be ignored completely?

rengaw83 commented:
The `$GLOBALS['TSFE']->rootLine` can't be removed @dkd-kaehm, see https://github.com/TYPO3-Solr/ext-solr/pull/3206#issuecomment-1046753551.  
How can we continue here to get this PR merged and records at pid 0 can be edited again?

rengaw83 commented:
Rebased and repushed, after conflicts from changes of main branch.
